lib/cloud/clients.go

```diff
@@ -942,3 +1029,108 @@ func (c *TestCloudClients) GetAzureRunCommandClient(subscription string) (azure.
 func (c *TestCloudClients) Close() error {
 	return nil
 }
+
+// AWSSessionCacheKeyBuilder is a helper struct wrapper around a string builder
+// that is used to efficiently build cache keys for a given region and, optionally,
+// a chain of AWS assumed-role sessions.
+type AWSSessionCacheKeyBuilder struct {
+	sb        *strings.Builder
+	roleCount int
+}
+
+// NewAWSSessionCacheKeyBuilder constructs a new cache key builder intialized
+// with a specific AWS region.
+func NewAWSSessionCacheKeyBuilder(region string) AWSSessionCacheKeyBuilder {
+	sb := strings.Builder{}
+	_, _ = sb.WriteString(region) // infalliable.
+	return AWSSessionCacheKeyBuilder{sb: &sb}
+}
+
+// AddRole adds an AWS IAM role to the cache key builder.
+func (c *AWSSessionCacheKeyBuilder) AddRole(awsRole services.AssumeRole) {
+	// Role ARN can potentially come from user input via --db-user for some
+	// protocols.
+	// Out of paranoia for crafted user input, I build the key in a way that
+	// will avoid fetching an AWS session from cache that the user should not
+	// have access to.
+	key := fmt.Sprintf(":Role[%d]:ARN[%v]:ExternalID[%v]", c.roleCount, awsRole.RoleARN, awsRole.ExternalID)
+	_, _ = c.sb.WriteString(key) // infalliable.
```

Comment:
typo

Reply:
removed cache key builder in fc800e15f75f7ed8f38b88b6aba00a756c66d144